Check Flink 1.17 compatibility

Run connector on Flink 1.17
Add build configuration for 1.17
Update Readme.md

In case of bigger issues with 1.15, drop the support for it.

I have run the tests locally at the latest main and see 
`[ERROR] Tests run: 257, Failures: 3, Errors: 15, Skipped: 0`

Inspired by[https://blog.jetbrains.com/idea/2020/08/migrating-from-junit-4-to-junit-5/]( https://blog.jetbrains.com/idea/2020/08/migrating-from-junit-4-to-junit-5/ ) I amended the pom file to use  junit5, this allows tests to run at 1.15, 1.16, 1.17, 1.18.0, 1.18.1

at 1.18.1 there are a couple more errors.

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -18,7 +18,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        flink: [ "1.15.0", "1.15.3", "1.16.1" ]
+        flink: [ "1.16.3", "1.17.2", "1.18.1"]
     steps:
       - uses: actions/checkout@v3
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## [Unreleased]
 
+### Fixed
+   
+    Moved junit support to junit 5, allowing junits to be run against flink 1.17 and 1.18.
+
 ## [0.12.0] - 2024-03-22
 
 ### Added

--- a/pom.xml
+++ b/pom.xml
@@ -68,7 +68,7 @@ under the License.
 
     <!-- IMPORTANT: If you update Flink, remember to update link to its docs in maven-javadoc-plugin <links>
          section, omitting the patch part (so for 1.15.0 use 1.15). -->
-    <flink.version>1.15.0</flink.version>
+    <flink.version>1.16.3</flink.version>
 
     <target.java.version>11</target.java.version>
     <scala.binary.version>2.12</scala.binary.version>
@@ -76,7 +76,9 @@ under the License.
     <maven.compiler.target>${target.java.version}</maven.compiler.target>
     <log4j.version>2.17.2</log4j.version>
     <lombok.version>1.18.22</lombok.version>
-    <junit.jupiter.version>5.8.1</junit.jupiter.version>
+    <junit4.version>4.13.2</junit4.version>
+    <junit5.version>5.10.1</junit5.version>
+    <junit.jupiter.version>${junit5.version}</junit.jupiter.version>
     <assertj.core.version>3.21.0</assertj.core.version>
     <mockito.version>4.0.0</mockito.version>
     <wiremock.version>2.27.2</wiremock.version>
@@ -167,6 +169,20 @@ under the License.
 
     <dependency>
       <groupId>org.apache.flink</groupId>
+      <artifactId>flink-test-utils</artifactId>
+      <version>${flink.version}</version>
+      <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.flink</groupId>
+            <artifactId>flink-table-test-utils</artifactId>
+            <version>${flink.version}</version>
+            <scope>test</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>org.apache.flink</groupId>
       <artifactId>flink-table-common</artifactId>
       <version>${flink.version}</version>
       <type>test-jar</type>
@@ -178,13 +194,6 @@ under the License.
       <artifactId>flink-connector-base</artifactId>
       <version>${flink.version}</version>
       <type>test-jar</type>
-      <scope>test</scope>
-    </dependency>
-
-    <dependency>
-      <groupId>org.apache.flink</groupId>
-      <artifactId>flink-table-planner_${scala.binary.version}</artifactId>
-      <version>${flink.version}</version>
       <scope>test</scope>
     </dependency>
 
@@ -233,9 +242,9 @@ under the License.
     </dependency>
 
     <dependency>
-      <groupId>org.junit.jupiter</groupId>
-      <artifactId>junit-jupiter-params</artifactId>
-      <version>${junit.jupiter.version}</version>
+      <groupId>org.junit.vintage</groupId>
+      <artifactId>junit-vintage-engine</artifactId>
+      <version>${junit5.version}</version>
       <scope>test</scope>
     </dependency>
 
@@ -323,6 +332,8 @@ under the License.
             <exclude>**/HttpLookupConnectorOptions.class</exclude>
             <exclude>**/Slf4jHttpPostRequestCallback.class</exclude>
             <exclude>**/SelfSignedTrustManager.class</exclude>
+            <exclude>org/apache/calcite**</exclude>
+            <exclude>org/apache/flink**</exclude>
           </excludes>
         </configuration>
         <executions>

--- a/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkConnectionTest.java
+++ b/src/test/java/com/getindata/connectors/http/internal/sink/HttpSinkConnectionTest.java
@@ -14,15 +14,19 @@ import com.github.tomakehurst.wiremock.http.Fault;
 import com.github.tomakehurst.wiremock.stubbing.ServeEvent;
 import org.apache.flink.configuration.ConfigConstants;
 import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.MetricOptions;
 import org.apache.flink.metrics.Counter;
 import org.apache.flink.metrics.Metric;
 import org.apache.flink.metrics.MetricConfig;
 import org.apache.flink.metrics.MetricGroup;
 import org.apache.flink.metrics.reporter.MetricReporter;
+import org.apache.flink.metrics.reporter.MetricReporterFactory;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.testutils.junit.extensions.ContextClassLoaderExtension;
 import org.junit.jupiter.api.AfterEach;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
 import static com.github.tomakehurst.wiremock.client.WireMock.*;
 import static com.github.tomakehurst.wiremock.stubbing.Scenario.STARTED;
 import static org.junit.jupiter.api.Assertions.assertEquals;
@@ -35,17 +39,26 @@ import com.getindata.connectors.http.internal.sink.httpclient.JavaNetSinkHttpCli
 
 public class HttpSinkConnectionTest {
 
+    @RegisterExtension
+    static final ContextClassLoaderExtension CONTEXT_CLASS_LOADER_EXTENSION =
+            ContextClassLoaderExtension.builder()
+                    .withServiceEntry(
+                            MetricReporterFactory.class,
+                            SendErrorsTestReporterFactory.class.getName())
+                    .build();
+
+
     private static final int SERVER_PORT = 9090;
 
     private static final int HTTPS_SERVER_PORT = 8443;
 
     private static final Set<Integer> messageIds = IntStream.range(0, 50)
-        .boxed()
-        .collect(Collectors.toSet());
+            .boxed()
+            .collect(Collectors.toSet());
 
     private static final List<String> messages = messageIds.stream()
-        .map(i -> "{\"http-sink-id\":" + i + "}")
-        .collect(Collectors.toList());
+            .map(i -> "{\"http-sink-id\":" + i + "}")
+            .collect(Collectors.toList());
 
     private StreamExecutionEnvironment env;
 
@@ -53,13 +66,15 @@ public class HttpSinkConnectionTest {
 
     @BeforeEach
     public void setUp() {
-        SendErrorsTestReporter.reset();
+        SendErrorsTestReporterFactory.reset();
 
         env = StreamExecutionEnvironment.getExecutionEnvironment(new Configuration() {
             {
-                this.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "sendErrorsTestReporter." +
-                        ConfigConstants.METRICS_REPORTER_CLASS_SUFFIX,
-                    SendErrorsTestReporter.class.getName());
+                setString(
+                        ConfigConstants.METRICS_REPORTER_PREFIX
+                                + "test."
+                                + MetricOptions.REPORTER_FACTORY_CLASS.key(),
+                        SendErrorsTestReporterFactory.class.getName());
             }
         });
 
@@ -85,7 +100,7 @@ public class HttpSinkConnectionTest {
         };
 
         List<Map<Object, Object>> responses =
-            testConnection(SinkRequestSubmitMode.SINGLE, responseMapper);
+                testConnection(SinkRequestSubmitMode.SINGLE, responseMapper);
 
         var idsSet = new HashSet<>(messageIds);
         for (var request : responses) {
@@ -104,14 +119,14 @@ public class HttpSinkConnectionTest {
         Function<ServeEvent, List<Map<Object, Object>>> responseMapper = response -> {
             try {
                 return new ObjectMapper().readValue(response.getRequest().getBody(),
-                    new TypeReference<List<Map<Object, Object>>>(){});
+                        new TypeReference<List<Map<Object, Object>>>(){});
             } catch (IOException e) {
                 throw new RuntimeException(e);
             }
         };
 
         List<List<Map<Object, Object>>> responses =
-            testConnection(SinkRequestSubmitMode.BATCH, responseMapper);
+                testConnection(SinkRequestSubmitMode.BATCH, responseMapper);
 
         var idsSet = new HashSet<>(messageIds);
         for (var requests : responses) {
@@ -134,39 +149,39 @@ public class HttpSinkConnectionTest {
         String contentTypeHeader = "application/json";
 
         wireMockServer.stubFor(any(urlPathEqualTo(endpoint))
-            .withHeader("Content-Type", equalTo(contentTypeHeader))
-            .willReturn(
-                aResponse().withHeader("Content-Type", contentTypeHeader)
-                    .withStatus(200)
-                    .withBody("{}")));
+                .withHeader("Content-Type", equalTo(contentTypeHeader))
+                .willReturn(
+                        aResponse().withHeader("Content-Type", contentTypeHeader)
+                                .withStatus(200)
+                                .withBody("{}")));
 
         var source = env.fromCollection(messages);
         var httpSink = HttpSink.<String>builder()
-            .setEndpointUrl("http://localhost:" + SERVER_PORT + endpoint)
-            .setElementConverter(
-                (s, _context) ->
+                .setEndpointUrl("http://localhost:" + SERVER_PORT + endpoint)
+                .setElementConverter(
+                    (s, _context) ->
                     new HttpSinkRequestEntry("POST", s.getBytes(StandardCharsets.UTF_8)))
-            .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
-            .setProperty(
-                HttpConnectorConfigConstants.SINK_HEADER_PREFIX + "Content-Type",
-                contentTypeHeader)
-            .setProperty(
-                HttpConnectorConfigConstants.SINK_HTTP_REQUEST_MODE,
-                mode.getMode()
-            )
-            .build();
+                .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
+                .setProperty(
+                        HttpConnectorConfigConstants.SINK_HEADER_PREFIX + "Content-Type",
+                        contentTypeHeader)
+                .setProperty(
+                        HttpConnectorConfigConstants.SINK_HTTP_REQUEST_MODE,
+                        mode.getMode()
+                )
+                .build();
         source.sinkTo(httpSink);
         env.execute("Http Sink test connection");
 
         var responses = wireMockServer.getAllServeEvents();
         assertTrue(responses.stream()
-            .allMatch(response -> Objects.equals(response.getRequest().getUrl(), endpoint)));
+                .allMatch(response -> Objects.equals(response.getRequest().getUrl(), endpoint)));
         assertTrue(
-            responses.stream().allMatch(response -> response.getResponse().getStatus() == 200));
+                responses.stream().allMatch(response -> response.getResponse().getStatus() == 200));
         assertTrue(responses.stream()
-            .allMatch(response -> Objects.equals(response.getRequest().getUrl(), endpoint)));
+                .allMatch(response -> Objects.equals(response.getRequest().getUrl(), endpoint)));
         assertTrue(
-            responses.stream().allMatch(response -> response.getResponse().getStatus() == 200));
+                responses.stream().allMatch(response -> response.getResponse().getStatus() == 200));
 
         List<T> collect = responses.stream().map(responseMapper).collect(Collectors.toList());
         assertTrue(collect.stream().allMatch(Objects::nonNull));
@@ -176,30 +191,30 @@ public class HttpSinkConnectionTest {
     @Test
     public void testServerErrorConnection() throws Exception {
         wireMockServer.stubFor(any(urlPathEqualTo("/myendpoint"))
-            .withHeader("Content-Type", equalTo("application/json"))
-            .inScenario("Retry Scenario")
-            .whenScenarioStateIs(STARTED)
-            .willReturn(serverError())
-            .willSetStateTo("Cause Success"));
+                .withHeader("Content-Type", equalTo("application/json"))
+                .inScenario("Retry Scenario")
+                .whenScenarioStateIs(STARTED)
+                .willReturn(serverError())
+                .willSetStateTo("Cause Success"));
         wireMockServer.stubFor(any(urlPathEqualTo("/myendpoint"))
-            .withHeader("Content-Type", equalTo("application/json"))
-            .inScenario("Retry Scenario")
-            .whenScenarioStateIs("Cause Success")
-            .willReturn(aResponse().withStatus(200))
-            .willSetStateTo("Cause Success"));
+                .withHeader("Content-Type", equalTo("application/json"))
+                .inScenario("Retry Scenario")
+                .whenScenarioStateIs("Cause Success")
+                .willReturn(aResponse().withStatus(200))
+                .willSetStateTo("Cause Success"));
 
         var source = env.fromCollection(List.of(messages.get(0)));
         var httpSink = HttpSink.<String>builder()
-            .setEndpointUrl("http://localhost:" + SERVER_PORT + "/myendpoint")
-            .setElementConverter(
-                (s, _context) ->
+                .setEndpointUrl("http://localhost:" + SERVER_PORT + "/myendpoint")
+                .setElementConverter(
+                    (s, _context) ->
                     new HttpSinkRequestEntry("POST", s.getBytes(StandardCharsets.UTF_8)))
-            .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
-            .build();
+                .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
+                .build();
         source.sinkTo(httpSink);
         env.execute("Http Sink test failed connection");
 
-        assertEquals(1, SendErrorsTestReporter.getCount());
+        assertEquals(1, SendErrorsTestReporterFactory.getCount());
         // TODO: reintroduce along with the retries
         //  var postedRequests = wireMockServer
         //  .findAll(postRequestedFor(urlPathEqualTo("/myendpoint")));
@@ -211,31 +226,32 @@ public class HttpSinkConnectionTest {
     @Test
     public void testFailedConnection() throws Exception {
         wireMockServer.stubFor(any(urlPathEqualTo("/myendpoint"))
-            .withHeader("Content-Type", equalTo("application/json"))
-            .inScenario("Retry Scenario")
-            .whenScenarioStateIs(STARTED)
-            .willReturn(aResponse().withFault(Fault.EMPTY_RESPONSE))
-            .willSetStateTo("Cause Success"));
+                .withHeader("Content-Type", equalTo("application/json"))
+                .inScenario("Retry Scenario")
+                .whenScenarioStateIs(STARTED)
+                .willReturn(aResponse().withFault(Fault.EMPTY_RESPONSE))
+                .willSetStateTo("Cause Success"));
 
         wireMockServer.stubFor(any(urlPathEqualTo("/myendpoint"))
-            .withHeader("Content-Type", equalTo("application/json"))
-            .inScenario("Retry Scenario")
-            .whenScenarioStateIs("Cause Success")
-            .willReturn(aResponse().withStatus(200))
-            .willSetStateTo("Cause Success"));
+                .withHeader("Content-Type", equalTo("application/json"))
+                .inScenario("Retry Scenario")
+                .whenScenarioStateIs("Cause Success")
+                .willReturn(aResponse().withStatus(200))
+                .willSetStateTo("Cause Success"));
 
         var source = env.fromCollection(List.of(messages.get(0)));
         var httpSink = HttpSink.<String>builder()
-            .setEndpointUrl("http://localhost:" + SERVER_PORT + "/myendpoint")
-            .setElementConverter(
-                (s, _context) ->
-                    new HttpSinkRequestEntry("POST", s.getBytes(StandardCharsets.UTF_8)))
-            .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
-            .build();
+                .setEndpointUrl("http://localhost:" + SERVER_PORT + "/myendpoint")
+                .setElementConverter(
+                    (s, _context) ->
+                    new HttpSinkRequestEntry("POST",
+                            s.getBytes(StandardCharsets.UTF_8)))
+                .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
+                .build();
         source.sinkTo(httpSink);
         env.execute("Http Sink test failed connection");
 
-        assertEquals(1, SendErrorsTestReporter.getCount());
+        assertEquals(1, SendErrorsTestReporterFactory.getCount());
         // var postedRequests = wireMockServer
         // .findAll(postRequestedFor(urlPathEqualTo("/myendpoint")));
         // assertEquals(2, postedRequests.size());
@@ -246,28 +262,28 @@ public class HttpSinkConnectionTest {
     @Test
     public void testFailedConnection404OnWhiteList() throws Exception {
         wireMockServer.stubFor(any(urlPathEqualTo("/myendpoint"))
-            .withHeader("Content-Type", equalTo("application/json"))
-            .willReturn(aResponse().withBody("404 body").withStatus(404)));
+                .withHeader("Content-Type", equalTo("application/json"))
+                .willReturn(aResponse().withBody("404 body").withStatus(404)));
 
         var source = env.fromCollection(List.of(messages.get(0)));
         var httpSink = HttpSink.<String>builder()
-            .setEndpointUrl("http://localhost:" + SERVER_PORT + "/myendpoint")
-            .setElementConverter(
-                (s, _context) ->
+                .setEndpointUrl("http://localhost:" + SERVER_PORT + "/myendpoint")
+                .setElementConverter(
+                    (s, _context) ->
                     new HttpSinkRequestEntry("POST", s.getBytes(StandardCharsets.UTF_8)))
-            .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
-            .setProperty("gid.connector.http.sink.error.code.exclude", "404, 405")
-            .setProperty("gid.connector.http.sink.error.code", "4XX")
-            .build();
+                .setSinkHttpClientBuilder(JavaNetSinkHttpClient::new)
+                .setProperty("gid.connector.http.sink.error.code.exclude", "404, 405")
+                .setProperty("gid.connector.http.sink.error.code", "4XX")
+                .build();
         source.sinkTo(httpSink);
         env.execute("Http Sink test failed connection");
 
-        assertEquals(0, SendErrorsTestReporter.getCount());
+        assertEquals(0, SendErrorsTestReporterFactory.getCount());
     }
 
     // must be public because of the reflection
-    public static class SendErrorsTestReporter implements MetricReporter {
-
+    public static class SendErrorsTestReporterFactory
+            implements MetricReporter, MetricReporterFactory {
         static volatile List<Counter> numRecordsSendErrors = null;
 
         public static long getCount() {
@@ -299,6 +315,11 @@ public class HttpSinkConnectionTest {
 
         @Override
         public void notifyOfRemovedMetric(Metric metric, String s, MetricGroup metricGroup) {
+        }
+
+        @Override
+        public MetricReporter createMetricReporter(Properties properties) {
+            return new SendErrorsTestReporterFactory();
         }
     }
 }

